### PR TITLE
[PATCH v6] api: ipsec: define more default values & improve SA info API

### PR DESCRIPTION
**Pull request by JannePeltonen**
Define default values for more fields of odp_ipsec_sa_param_t so that applications do not always have to explicitly initialize as many fields after the odp_ipsec_sa_param_init() call.

The current validation tests do not actually set all the parameters that they should. It can cause test failures in implementations that set the implementation-defined values differently from linux-gen. This API change makes the needed fix there smaller.

### CODE REVIEW

**Comment by JannePeltonen** [comment]
v2:
- Change some of the unions inside odp_ipsec_sa_param_t to structs so that it is possible to define (nonzero) defaults that the init function can set. This may also help in backward compatibility of future API changes as new fields with non-zero defaults can be added in all parts of the parameter structure.
- Add linux-gen implementation

**Comment by JereLeppanen** [comment]
Typo in the commit comment of the first commit: "paramteter".

File: include/odp/api/spec/ipsec.h [hunk 1]
```diff
@@ -712,10 +714,12 @@ typedef struct odp_ipsec_sa_param_t {
 	uint32_t context_len;
 
 	/** IPSEC SA direction dependent parameters */
-	union {
+	struct {
```
Comment (anoobj) [inline include/odp/api/spec/ipsec.h:717]: Is this intentional?

File: test/validation/api/ipsec/ipsec_test_out.c [hunk 2]
```diff
@@ -1371,6 +1371,53 @@ static void ipsec_test_capability(void)
...
+	CU_ASSERT(config.inbound.lookup.spi_overlap == 0);
+	CU_ASSERT(config.inbound.retain_outer == ODP_PROTO_LAYER_NONE);
+	CU_ASSERT(config.inbound.parse_level == ODP_PROTO_LAYER_NONE);
+	CU_ASSERT(config.inbound.chksums.all_chksum == 0);
+	CU_ASSERT(config.outbound.all_chksum == 0);
+	CU_ASSERT(!config.stats_en);
+
+	odp_ipsec_sa_param_init(&sa_param);
+	CU_ASSERT(sa_param.proto == ODP_IPSEC_ESP);
+	CU_ASSERT(sa_param.crypto.cipher_alg == ODP_CIPHER_ALG_NULL);
```
Comment (anoobj) [inline test/validation/api/ipsec/ipsec_test_out.c:1394]: Just wondering what should be the behavior if the implementation doesn't support NULL algos.

File: include/odp/api/spec/ipsec.h (hunk 1 above)
Comment (JannePeltonen) [reply include/odp/api/spec/ipsec.h:717]: Yes it is, as explained in the github comment and the commit message.

Using a union has the drawback that default values cannot be defined for the union members as odp_ipsec_sa_param_init() cannot write them without affecting all the union members at the same time. And not being able to define defaults makes the _param_init() pattern used in ODP less useful. Changing the union to a struct fixes this.

Increased memory usage is a drawback but maybe it does not matter too much since this structure is needed only at SA creation time for parameter passing. In my x86 build the size seems to increase from 208 bytes to 272 bytes.

So, does the change make sense or should we just stick with the unions (and define fewer default values)?

File: test/validation/api/ipsec/ipsec_test_out.c (hunk 2 above)
Comment (JannePeltonen) [reply test/validation/api/ipsec/ipsec_test_out.c:1394]: TL; DR: We can leave things as they are or consider changing the default authentication algorithm to SHA-1.

odp_ipsec_sa_param_init() is supposed to set the default values as specified in the API spec regardless of the actual algorithm support in the implementation. As you must have noticed, this PR does not change the default algorithms but just adds a test for what was already required.

If an implementation does not support null encryption or null authentication, we have a situation where the default parameters set by odp_ipsec_sa_param_init() cannot be passed to odp_ipsec_sa_create() and the application has to set the parameters to suitable values. But that is not much different from parameters which do not have default value defined.

That said, null encryption in ESP is likely supported by many implementations because it is mandatory in IPsec and also tends to be somewhat easier to implement than most other ciphers.

Null authentication is no longer mandatory in IPsec and even if it is supported, it may not be combined with null encryption in real usage (but in performance tuning such combo can be useful and ODP API allows it).

Maybe SHA-1, which is mandatory in IPsec and thus likely supported by most implementations, would be a better default value for the authentication algorithm than the current null authentication.

File: include/odp/api/spec/ipsec.h (hunk 1 above)
Comment (anoobj) [reply include/odp/api/spec/ipsec.h:717]: The only drawback is, sa_info is supposed to return this struct and the memory requirement would just increase. Or if we can state that odp_ipsec_sa_param_t.inbound/outbound would be ignored based on odp_ipsec_sa_param_t.dir, then it should be fine.

IOW, when dir = inbound, sa_info.param.outbound would have default values (& the other way round also). Would that be fine?

Comment (JannePeltonen) [reply include/odp/api/spec/ipsec.h:717]: I think the SA info API is a bit problematic as the current wording suggests that the returned param struct is a exact (?) copy of the one used in SA creation. But we do not want that as we do not want to force implementations to store the struct or the values of the ignored fields that cannot be recreated from internal SA data.

This problem exists already before the union-to-struct change, but the change makes it words.

I think I would like to propose the following change to the sa_info API, regardless of what we do with this PR:

```
 typedef struct odp_ipsec_sa_info_t {
-       /** Copy of IPSEC Security Association (SA) parameters */
+       /** IPsec SA parameters
+        *
+        * This is not necessarily an exact copy of the actual parameter
+        * structure used in SA creation. The fields that were relevant
+        * for the SA in the creation phase will have the same values,
+        * but other fields, such as tunnel parameters for a transport
+        * mode SA, will have undefined values.
+        */
        odp_ipsec_sa_param_t param;
```

Comment (anoobj) [reply include/odp/api/spec/ipsec.h:717]: I agree to the suggestion on sa_info. With that change, move to struct is not an issue. Agreed that separate structs allows for specifying defaults better.

**Comment by JannePeltonen** [comment]
v3:
- Added SA info API change to relax the requirements toward the returned parameter struct
- Fixed the typo in a commit message
- Rebased

File: include/odp/api/spec/ipsec.h [hunk 3]
```diff
@@ -765,7 +769,9 @@ typedef struct odp_ipsec_sa_param_t {
 			/** Parameters for tunnel mode */
 			odp_ipsec_tunnel_param_t tunnel;
 
-			/** Fragmentation mode */
+			/** Fragmentation mode
+			 *  The default value is ODP_IPSEC_RRAG_DISABLED.
```
Comment (psavol) [inline include/odp/api/spec/ipsec.h:773]: Typo: ODP_IPSEC_RRAG_DISABLED -> FRAG

**Comment by psavol** [comment]
First commit log entry has typo: ... remember the content of also those fields thae -> that

File: test/validation/api/ipsec/ipsec_test_out.c (hunk 2 above)
Comment (anoobj) [reply test/validation/api/ipsec/ipsec_test_out.c:1394]: > If an implementation does not support null encryption or null authentication, we have a situation where the default 
> parameters set by odp_ipsec_sa_param_init() cannot be passed to odp_ipsec_sa_create() and the application has to set
> the parameters to suitable values.

This is what I had in mind. If application is expected to check crypto capabilities before doing odp_ipsec_sa_create(), then we can keep existing behavior. RFC compliance would be the rationale to state algos as mandatory. But implementations need not be RFC compliant. 

Other option is to leave default values for algos as implementation defined. Would that be fine? But then testing such a field would be very complicated.

**Comment by JannePeltonen** [comment]
v4: Fixed the two typos.

File: test/validation/api/ipsec/ipsec_test_out.c (hunk 2 above)
Comment (JannePeltonen) [reply test/validation/api/ipsec/ipsec_test_out.c:1394]: > If application is expected to check crypto capabilities before doing odp_ipsec_sa_create()

Yes, it is expected to do that or accept that the creation may fail.

> Other option is to leave default values for algos as implementation defined. Would that be fine?

I think that would be ok too, but I do not think it would be any more useful to an application than the null defaults. You cannot just let the implementation choose the crypto algorithms for an SA and you would have to give keys of correct length too.

Note that this PR is not changing the default values for the algorithms but only adds a validation check for what was already specified in the API.

Comment (anoobj) [reply test/validation/api/ipsec/ipsec_test_out.c:1394]: > You cannot just let the implementation choose the crypto algorithms for an SA and you would have to give keys of
> correct length too

Yeah. In that sense, it is better to retain the current behavior (with NULLs).

> Note that this PR is not changing the default values for the algorithms but only adds a validation check for what was
> already specified in the API.

Agreed. I was just trying to understand the spec requirements in case NULL algos are not supported.

File: include/odp/api/spec/ipsec.h [hunk 4]
```diff
@@ -491,11 +491,11 @@ typedef struct odp_ipsec_ipv6_param_t {
...
  */
 typedef struct odp_ipsec_tunnel_param_t {
-	/** Tunnel type: IPv4 or IPv6 */
+	/** Tunnel type: IPv4 or IPv6. The default is IPv4. */
 	odp_ipsec_tunnel_type_t type;
 
-	/** Variant mappings for tunnel parameters */
-	union {
+	/** Tunnel type specific parameters */
+	struct {
```
Comment (anoobj) [inline include/odp/api/spec/ipsec.h:498]: Missed this earlier. Is this change required? Or are we replacing all unions with struct in general? We are not specifying default values for these, right?

Comment (psavol) [reply include/odp/api/spec/ipsec.h:498]: Defaults are defined at odp_ipsec_ipv4_param_t  and odp_ipsec_ipv6_param_t definition e.g. ttl and hlimit.

Comment (anoobj) [reply include/odp/api/spec/ipsec.h:498]: Missed it. Thanks!

**Review by psavol — APPROVED** [review]

**Comment by anoobj** [comment]
Reviewed-by: Anoob Joseph <anoobj@marvell.com>

**Comment by JannePeltonen** [comment]
v5: Added review tags
